safekeeper/src/wal_backup.rs

```diff
@@ -0,0 +1,267 @@
+use std::path::PathBuf;
+
+use postgres_ffi::xlog_utils::{XLogSegNo, PG_TLI, XLogFileName, XLogSegNoOffsetToRecPtr};
+use tokio::fs::File;
+use tokio::runtime::{Builder, Runtime};
+
+use anyhow::{Result, ensure, bail, Context};
+
+use tokio::sync::watch::{Receiver, Sender};
+use utils::{
+    lsn::Lsn,
+    zid::{ZTenantTimelineId},
+};
+
+use tracing::*;
+
+use crate::config::{RemoteStorageConfig, RemoteStorageKind};
+use crate::remote_storage::{NoopStorage, RemoteStorage, LocalFs, S3Bucket};
+use crate::{broker, SafeKeeperConf, wal_storage};
+
+
+use once_cell::sync::OnceCell;
+
+static BACKUP_RUNTIME: OnceCell<Runtime> = OnceCell::new();
+const DEFAULT_BACKUP_RUNTIME_SIZE: u32 = 16;
+
+const MIN_WAL_SEGMENT_SIZE: usize = 1 << 20; // 1MB
+const MAX_WAL_SEGMENT_SIZE: usize = 1 << 30; // 1GB
+const DEFAULT_WAL_SEGMENT_SIZE: usize = 16 << 20; // 16MB
+
+const BACKUP_ELECTION_PATH: &str = "WAL_BACKUP";
+
+async fn detect_task(
+    conf: SafeKeeperConf,
+    timeline_id: ZTenantTimelineId,
+    backup_start: Lsn,
+    mut wal_seg_size: Receiver<u32>,
+    lease_id : Option<i64>, 
+    mut lsn_durable: Receiver<Lsn>, 
+    lsn_backed_up: Sender<Lsn>,
+    ) -> Result<()> {
+
+    // TODO: antons come up with better naming
+    wal_seg_size.changed().await.expect("Failed to recieve wal segment size");
+    let wal_seg_size = *wal_seg_size.borrow();
+    let mut wal_seg_size = usize::try_from(wal_seg_size).with_context(|| format!("Couldn't convert {} to usize", wal_seg_size))?;
+    if !(MIN_WAL_SEGMENT_SIZE..=MAX_WAL_SEGMENT_SIZE).contains(&wal_seg_size) {
+        warn!("Invalid wal seg size provided, should be at least 1Mb per postgres, setting to default value of {}", DEFAULT_WAL_SEGMENT_SIZE);
+        wal_seg_size = DEFAULT_WAL_SEGMENT_SIZE; 
+    }
+
+    let mut backup_lsn = backup_start;
+
+    loop {
+        if let Some(lease_id) = lease_id {
+            broker::become_leader(lease_id, BACKUP_ELECTION_PATH.to_string(), &timeline_id, &conf).await.ok();
+        }
+
+        while lsn_durable.changed().await.is_ok() {
+            let commit_lsn = *lsn_durable.borrow();
+
+            if let Some(lease_id) = lease_id {
+                if broker::is_leader(lease_id, BACKUP_ELECTION_PATH.to_string(), &timeline_id, &conf).await.is_err() {
+                    info!("Leadership is lost for timeline {}", timeline_id);
+                    break;
+                }
+            }
+
+            if commit_lsn < backup_lsn {
+                bail!("Commit lsn {} moved back {}", commit_lsn, backup_lsn);
+            }
+
+            if backup_lsn.segment_number(wal_seg_size) == commit_lsn.segment_number(wal_seg_size) {
+                continue;
```

Comment:
Curious: should we check that we have something to backup before contending for leadership?

I would also keep a single client all the time, without creating and connecting every time I need to adjust the leadership: I wonder, if dealing with leadership even supposed to used this way with reconnects?

Reply:
Yes, now that I realise that we have `lease_keep_alive` with another client polling the whole lease liveness, I wonder how this all is supposed to work.

I would create the client here, polling for both the lease, leadership and new events in the same loop, cancelling the loop if no lease is granted, doing nothing if not a leader, etc?

Reply:
functionally the usage of a new client is okay for leadership validation is fine as the leader is associated with a lease (and once the lease expires a new candidate would become a leader if such a candidate exists). Currently caching a client makes code more complicated as the client needs to be mutable while participating in multiple activities. I would defer optimization of this code path until we get a better idea if we want to have general timeline leaders not backup specific.

_note: agree that flipping lsn check with leadership check is a freebie that would improve preformance_